web/components/contract-search.tsx

```diff
@@ -167,80 +169,67 @@ export function ContractSearch(props: {
-
-    algoliaIndex
-      .search(query, {
+  const [pages, setPages] = useState<Contract[][]>([])
+  const requestId = useRef(0)
+
+  const queryNextPage = async () => {
+    const id = ++requestId.current
+    if (pages.length < numPages) {
+      const algoliaIndex = query ? searchIndex : index
+      const results = await algoliaIndex.search(query, {
         facetFilters,
         numericFilters,
-        page,
+        page: pages.length,
         hitsPerPage: 20,
       })
-      .then((results) => {
-        if (!wasMostRecentQuery) return
-
-        if (page === 0) {
-          setHitsByPage({
-            [0]: results.hits as any as Contract[],
-          })
-        } else {
-          setHitsByPage((hitsByPage) => ({
-            ...hitsByPage,
-            [page]: results.hits,
-          }))
-        }
-        setNumPages(results.nbPages)
-      })
-    return () => {
-      wasMostRecentQuery = false
+      // if there's a more recent request, forget about this one
+      if (id === requestId.current) {
+        // this spooky looking function is the easiest way to get react to
+        // batch this and not do two renders. we can throw it out in react 18.
+        // see https://github.com/reactwg/react-18/discussions/21
+        unstable_batchedUpdates(() => {
+          setPages((pages) => [...pages, results.hits as any as Contract[]])
+          setNumPages(results.nbPages)
+        })
+      }
     }
-    // Note numeric filters are unique based on current time, so can't compare
-    // them by value.
-  }, [query, page, index, searchIndex, JSON.stringify(facetFilters), filter])
-
-  const loadMore = () => {
-    if (page >= numPages - 1) return
-
-    const haveLoadedCurrentPage = hitsByPage[page]
-    if (haveLoadedCurrentPage) setPage(page + 1)
   }
 
-  const hits = range(0, page + 1)
-    .map((p) => hitsByPage[p] ?? [])
-    .flat()
+  useEffect(() => {
+    // if there are no search results yet (not even an empty page), that means
+    // we should do an initial search to populate a first page.
+    if (pages.length === 0) {
+      queryNextPage()
+    }
+  }, [pages])
 
-  const contracts = hits.filter(
-    (c) => !additionalFilter?.excludeContractIds?.includes(c.id)
-  )
+  const contracts = pages
+    .flat()
+    .filter((c) => !additionalFilter?.excludeContractIds?.includes(c.id))
 
   const showTime =
     sort === 'close-date' || sort === 'resolve-date' ? sort : undefined
 
   const updateQuery = (newQuery: string) => {
     setQuery(newQuery)
-    setPage(0)
+    setPages([])
+    setNumPages(1)
   }
 
   const selectFilter = (newFilter: filter) => {
     if (newFilter === filter) return
     setFilter(newFilter)
-    setPage(0)
+    setPages([])
+    setNumPages(1)
     track('select search filter', { filter: newFilter })
   }
 
   const selectSort = (newSort: Sort) => {
     if (newSort === sort) return
-
-    setPage(0)
     setSort(newSort)
+    setPages([])
```

Comment:
Setting pages to empty here would mean that while the next search is happening, we'd flash empty results (or a loading indicator), right?

The previous behavior is to keep showing the old search until the next search is loaded. It's less jumpy for the user.

Ditto for `updateQuery` and `selectFilter`

Reply:
Oh, I actually didn't realize the old behavior was different. I personally like the new behavior, but I don't have a strong preference, so if you like the old behavior, I will change it back.